kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -194,9 +195,7 @@ private class SemaphoreSegment(id: Long, prev: SemaphoreSegment?): Segment<Semap
 
     // Cleans the acquirer slot located by the specified index
     // and removes this segment physically if all slots are cleaned.
-    fun cancel(index: Int) {
-        // Clean the specified waiter
-        acquirers[index].value = CANCELLED
+    fun cancel(index: Int): Boolean = getAndUpdate(index, CANCELLED) !== RESUMED.also {
```

Comment:
With `also { }` this function now has the following order of operations:
1.  `cancelledSlots.incrementAndGet()` & `remove()`
2. `getAndUpdate(index, CANCELLED)`

Is it a bug (need a test to check it) or is it also Ok to do in this order? 
Anyway, please rewrite in a more explicit way.

Reply:
I think it works, but I intended to perform `getAndUpdate` at first; otherwise, it is very complicated to prove that the code is correct.